src/lib.rs

```diff
@@ -5443,14 +5442,36 @@ pub unsafe trait Unaligned {
         Self: Sized;
 }
 
-/// Derives an optimized implementation of [`Hash`] for types that implement
-/// [`IntoBytes`] and [`Immutable`].
+/// Derives an optimized [`Hash`] implementation.
 ///
-/// The standard library's derive for `Hash` generates a recursive descent
-/// into the fields of the type it is applied to. Instead, the implementation
-/// derived by this macro makes a single call to [`Hasher::write()`] for both
-/// [`Hash::hash()`] and [`Hash::hash_slice()`], feeding the hasher the bytes
-/// of the type or slice all at once.
+/// This derive can be applied to structs and enums implementing both
+/// [`Immutable`] and [`IntoBytes`]; e.g.:
+///
+/// ```
+/// # use zerocopy_derive::{ByteHash, Immutable, IntoBytes};
+/// #[derive(ByteHash, Immutable, IntoBytes)]
+/// #[repr(C)]
+/// struct MyStruct {
+/// # /*
+///     ...
+/// # */
+/// }
+///
+/// #[derive(ByteHash, Immutable, IntoBytes)]
+/// #[repr(u8)]
+/// enum MyEnum {
+/// #   Variant,
+/// # /*
+///     ...
+/// # */
+/// }
+/// ```
+///
+/// The standard library's [`derive(Hash)`][derive@Hash] produces hashes by
+/// individually hashing each field and combining the results. Instead, the
+/// implementations of [`Hash::hash()`] and [`Hash::hash_slice()`] generated by
+/// `derive(ByteHash)` convert the entirey of `self` to a byte slice and hashes
+/// it in a single call to [`Hasher::write()`].
```

Comment:
Maybe mention something about how this is often faster or easier to optimize or something to that effect? After all, that's the only reason you'd use this derive. Currently the only mention of optimization or performance in this doc comment is in the summary sentence.

Reply:
The "only mention" being in the summary sentence is fairly prominent positioning. I'm hesitant to over-promise here, because this derive *isn't* necessarily a performance slam dunk. For example, comparing two `u64`s for equality is likely to optimize better than casting them to slices of dynamic length and comparing them byte-by-byte. I'm really not sure where the inflection point is. 

For that reason, I think the honest thing to do is to explain the mechanics, but not make specific promises about performance.

Reply:
https://github.com/google/zerocopy/compare/f10667ea4dbd1a8d7802c1bbf09ba3d23bc81b08..a9faae2c1b9b5e45a02d52c24a9e984b44aa12b4